django/core/handlers/asgi.py

```diff
@@ -178,38 +178,44 @@ async def handle(self, scope, receive, send):
             body_file = await self.read_body(receive)
         except RequestAborted:
             return
-        # Request is complete and can be served.
-        set_script_prefix(get_script_prefix(scope))
-        await signals.request_started.asend(sender=self.__class__, scope=scope)
-        # Get the request and check for basic issues.
-        request, error_response = self.create_request(scope, body_file)
-        if request is None:
-            body_file.close()
-            await self.send_response(error_response, send)
-            return
-        # Try to catch a disconnect while getting response.
-        tasks = [
-            asyncio.create_task(self.run_get_response(request)),
-            asyncio.create_task(self.listen_for_disconnect(receive)),
-        ]
-        done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
-        done, pending = done.pop(), pending.pop()
-        # Allow views to handle cancellation.
-        pending.cancel()
-        try:
-            await pending
-        except asyncio.CancelledError:
-            # Task re-raised the CancelledError as expected.
-            pass
-        try:
-            response = done.result()
-        except RequestAborted:
-            body_file.close()
-            return
-        except AssertionError:
-            body_file.close()
-            raise
-        # Send the response.
+        with body_file:
+            # Request is complete and can be served.
+            set_script_prefix(get_script_prefix(scope))
+            await signals.request_started.asend(sender=self.__class__, scope=scope)
+            # Get the request and check for basic issues.
+            request, error_response = self.create_request(scope, body_file)
+            if request is None:
+                await self.send_response(error_response, send)
+                return
+
+            # Try to catch a disconnect while getting response.
+            tasks = [
+                # We will check the status of these tasks and (optionally)
+                # terminate them in this order. The listen_for_disconnect task
+                # goes first because it should not raise unexpected errors that
+                # would prevent us from cancelling get_and_send_response.
+                asyncio.create_task(self.listen_for_disconnect(receive)),
+                asyncio.create_task(self.get_and_send_response(request, send)),
+            ]
+            await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
+            # Now wait on both tasks (they may have both finished by now).
```

Comment:
@qxcv Can I ask what motivates this change? (Thanks!)

Reply:
Assuming you mean the change from

```
done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
done, pending = done.pop(), pending.pop()
# some conditionals to handle done and pending
```

to

```
await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
# a loop through tasks
```

The answer is that both tasks can finish at the same time, which means that both tasks end up in `done` and the `pending.pop()` errors out. I only observed this behavior when serving an app with Daphne, not in the existing unit tests. I tried to write a test for it last week but my asyncio-foo wasn't up to the task of forcing both tasks to finish at once. I'll try again later this week (I found [this great overview of asyncio](https://tenthousandmeters.com/blog/python-behind-the-scenes-12-how-asyncawait-works-in-python/) that cleared up a lot of things for me).

Reply:
> The answer is that both tasks can finish at the same time, which means that both tasks end up in done and the pending.pop() errors out. I only observed this behavior when serving an app with Daphne, not in the existing unit tests. I tried to write a test for it last week but my asyncio-foo wasn't up to the task of forcing both tasks to finish at once.

OK... that's really interesting 🤔 Since only a single task gets run on the event loop at a time, how they finish simultaneously is a curious… — but **yes** if that's happening that change make sense. 

If you can't come up with a test, maybe a minimal project that demonstrates would be good to see (if you can 🙏)

Reply:
I think this is resolved by the discussion below https://github.com/django/django/pull/17147#issuecomment-1689393302 

If someone has a strategy to test it then super, but checking both tasks (in a loop) can't hurt... — it's just that the initial implementation **assumed** one task would be still pending.